### PR TITLE
Previous quarter by period number

### PR DESCRIPTION
**Pull request by butofleury**

### CODE REVIEW

File: src/support/DatePeriods.js [hunk 1]
```diff
@@ -498,6 +498,20 @@ class DatePeriods {
       this.dhis2MonthPeriod(year, month)
     );
   }
+
+  static previousQuarterByPeriodNumber(period, previousPeriods){
+    var previousQuarter = "";
+    var arrayPreviousQuarter = [];
+    for(var i=0; i < previousPeriods; i++){
+      if(i === 0){
+        previousQuarter = this.previousQuarter(period);
```
Comment (mestachs) [inline src/support/DatePeriods.js]: It looks "dirty" in the DatePeriods and we have already a way to ask the "next" and "previous" period
https://github.com/BLSQ/blsq-report-components/blob/39ad3b5b33e60d30a1e943240e9111b0490878b5/src/support/DatePeriods.js#L227
https://github.com/BLSQ/blsq-report-components/blob/39ad3b5b33e60d30a1e943240e9111b0490878b5/src/support/DatePeriods.js#L243
I would reuse that instead calculating quarters and calling previousQuarter

I would not include the current period in the return array of periods
```
static previousPeriods(period, number_of_periods)
static nextPeriods(period, number_of_periods)
```

and add a test for all periods

```
previousPeriods("2018Q3", 3) == ["2017Q4", "2018Q1", "2018Q2"]
previousPeriods("201802", 3) == ["201711", "201712", "201801"]
previousPeriods("2018", 3) == ["2015","2016","2017"]
```
idem for `nextPeriods`

File: src/support/Dhis2.js [hunk 2]
```diff
@@ -312,8 +312,8 @@ class Dhis2 {
       orgUnits: orgUnits,
       period: period,
       quarterPeriod: period,
-      quarterPeriods: DatePeriods.split(period, "quarterly"),
-      monthlyPeriods: DatePeriods.split(period, "monthly"),
+      quarterPeriods: invoiceType.previousPeriods !== undefined ? DatePeriods.split(DatePeriods.previousQuarterByPeriodNumber(period, invoiceType.previousPeriods), "quarterly") : DatePeriods.split(period, "quarterly"),
```
Comment (mestachs) [inline src/support/Dhis2.js]: I would extract quarterPeriods and monthlyPeriods
in a variable and leave the  return as it was before.

```
  let quarterPeriods = DatePeriods.split(period, "quarterly")
  let monthlyPeriods = DatePeriods.split(period, "monthly")
  if (invoiceType.previousPeriods) {
    quarterPeriods = quarterPeriods.concat(DatePeriods.previousPeriods(DatePeriods.split(period, "quarterly")[0], invoiceType.previousPeriods))
    monthlyPeriods = monthlyPeriods.concat(DatePeriods.previousPeriods(DatePeriods.split(period, "monthly")[0], invoiceType.previousPeriods))
  }
       
```

IMO to be consistent you need to add yearlyPeriods too

**Comment by mestachs** [comment]
can you also modify one of the invoice in the example app to use that new feature

File: src/support/Dhis2.js [hunk 3]
```diff
@@ -307,13 +307,24 @@ class Dhis2 {
     const year = period.slice(0, 4);
     const quarter = DatePeriods.split(period, "quarterly")[0].slice(5, 6);
 
+    let quarterPeriods = DatePeriods.split(period, "quarterly");
+    let monthlyPeriods = DatePeriods.split(period, "monthly");
+    let yearlyPeriods =  DatePeriods.split(period, "yearly");
+
+    if (invoiceType.previousPeriods) {
+      quarterPeriods = quarterPeriods.concat(DatePeriods.previousPeriods(DatePeriods.split(period, "quarterly")[0], invoiceType.previousPeriods));
+      monthlyPeriods = monthlyPeriods.concat(DatePeriods.previousPeriods(DatePeriods.split(period, "monthly")[0], 3));
```
Comment (butofleury) [inline src/support/Dhis2.js]: I did it so to include all 3 months of the previous period!

**Pull request merged**